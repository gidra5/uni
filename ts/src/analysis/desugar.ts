import { node, NodeType, Tree } from "../ast";
import { assert } from "../utils";

export const desugar = (ast: Tree): Tree => {
  ast.children = ast.children.map(desugar);
  switch (ast.type) {
    case NodeType.PARENS:
      assert(ast.children.length === 1);
      if (ast.children[0].type === NodeType.IMPLICIT_PLACEHOLDER) return node(NodeType.UNIT);
      return ast.children[0];
    case NodeType.SCRIPT:
    case NodeType.MODULE:
    default:
      return ast;
  }
};
