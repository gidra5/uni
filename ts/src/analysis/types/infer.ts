import { Iterator } from "iterator-js";
import { NodeType, Tree } from "../../ast";
import { assert, nextId, unreachable } from "../../utils";
import { inject, Injectable } from "../../utils/injector";
import { UnificationTable } from "./unification";
import { PhysicalType, Type } from "./utils";
import { Binding, resolve } from "../scope";

export type TypeSchema = Map<number, Type>;
export type PhysicalTypeSchema = Map<number, PhysicalType>;

// subtype infers types that values must always satisfy (the actual values will be at least of these types)
// for example, if a parameter is only used as int (passed to a function that only accepts ints),
// then the type for the parameter will be int
// its value must inhabit the type
// if nothing can be inferred, then the type will be "unknown"

// supertype infers types that values will always satisfy (the actual values will be at most of these types)
// for example, if a function only called with ints, then the type for its arguments will be int

const scope = [
  ["print_int", { fn: { arg: "int", return: "int" } }],
  ["print_string", { fn: { arg: "string", return: "string" } }],
  ["print_float", { fn: { arg: "float", return: "float" } }],
  ["print_symbol", { fn: { arg: "symbol", return: "symbol" } }],
  ["true", "boolean"],
  ["false", "boolean"],
] as const satisfies [string, Type][];

export const globalResolvedNames = Iterator.iter(scope)
  .map<Binding>(([name]) => [name, nextId()])
  .toMap();
export const globalResolvedNamesArray = [...globalResolvedNames.entries()];

const globalNames = new Map<number, Type>(scope.map(([name, type]) => [globalResolvedNames.get(name)!, type]));

export class Context {
  unificationTable = new UnificationTable();
  names: Map<number, Type> = globalNames;
  constructor() {}

  bind(ast: Tree, f: (argType: Type) => Type): Type {
    const argType = inferPattern(ast);
    const name = inject(Injectable.NodeToVariableMap).get(ast.children[0].id)!;
    this.names.set(name, argType);
    const t = f(argType);
    this.names.delete(name);
    return t;
  }
}

// is `a` a subtype of `b`? `a <: b` == true
// the bottom type must always be a subtype of the top type
export const isSubtype = (a: Type, b: Type): boolean => {
  // console.log(a, b);
  if (a === b) return true;

  if (typeof b === "object" && "not" in b) return !isSubtype(b.not, a);
  if (typeof a === "object" && "not" in a) return !isSubtype(b, a.not);

  if (a === "void") return true;
  if (b === "unknown") return true;

  if (typeof a === "object" && "and" in a) {
    return a.and.some((type) => isSubtype(type, b));
    // // console.log(a, b);
    // const x = a.and.map((type) => isSubtype(type, b));
    // // console.log(3, x);
    // return x.some((x) => x);
  }
  if (typeof b === "object" && "and" in b) {
    return b.and.every((type) => isSubtype(a, type));
    // // console.log(a, b);
    // const x = b.and.map((type) => isSubtype(a, type));
    // // console.log(4, x);
    // return x.every((x) => x);
  }

  if (typeof a === "object" && "or" in a) {
    return a.or.every((type) => isSubtype(type, b));
    // console.log(a, b);
    // const x = a.or.map((type) => isSubtype(type, b));
    // console.log(2, x);
    // return x.every((x) => x);
  }
  if (typeof b === "object" && "or" in b) {
    return b.or.some((type) => isSubtype(a, type));
    // console.log(a, b);
    // const x = b.or.map((type) => isSubtype(a, type));
    // console.log(1, x);

    // return x.some((x) => x);
  }

  if (typeof a === "object" && "fn" in a) {
    if (!(typeof b === "object" && "fn" in b)) return false;
    return isSubtype(b.fn.arg, a.fn.arg) && isSubtype(a.fn.return, b.fn.return);
  }
  if (typeof b === "object" && "fn" in b) return false;

  if (typeof a === "object" && "variable" in a) {
    if (typeof b === "object" && "variable" in b) return a.variable === b.variable;
    return false;
  }
  if (typeof b === "object" && "variable" in b) return false;

  if (typeof a === "object" && "record" in a) {
    if (!(typeof b === "object" && "record" in b)) return false;
    if (b.record.length === 0) return true;
    if (b.record.length > a.record.length) return false;
    if (a.labels) {
      if (!b.labels) return false;
      const bMap = Iterator.iter(b.labels).zip(b.record).toMap();
      return Iterator.iter(a.labels)
        .zip(a.record)
        .every(([label, type]) => bMap.has(label) && isSubtype(type, bMap.get(label)!));
    }
    return Iterator.iter(a.record)
      .zip(b.record)
      .every(([a, b]) => isSubtype(a, b));
  }

  return false;
};

export const isSubtypeEqual = (a: Type, b: Type): boolean => isSubtype(a, b) && isSubtype(b, a);

const inferPattern = (a: Tree): Type => {
  switch (a.type) {
    case NodeType.NAME:
      if (a.data.value === "void") return "void";
      if (a.data.value === "unknown") return "unknown";
      if (a.data.value === "int") return "int";
      if (a.data.value === "float") return "float";
      if (a.data.value === "string") return "string";
      if (a.data.value === "bool") return "boolean";
      return { variable: a.id };
    case NodeType.LABEL:
      return inferPattern(a.children[1]);
    case NodeType.FUNCTION:
      return { fn: { arg: inferPattern(a.children[0]), return: inferPattern(a.children[1]) } };
    default:
      unreachable("cant infer type");
  }
};

export const infer = (ast: Tree, context: Context): Type => {
  console.dir({ log: 1, context, ast }, { depth: null });

  switch (ast.type) {
    case NodeType.NUMBER: {
      const type = Number.isInteger(ast.data.value) ? "int" : "float";
      context.unificationTable.addConstraint(ast.id, { exactly: type });
      return type;
    }
    case NodeType.STRING:
      context.unificationTable.addConstraint(ast.id, { exactly: "string" });
      return "string";
    case NodeType.NAME: {
      const name = inject(Injectable.NodeToVariableMap).get(ast.id)!;
      const type = context.names.get(name);

      if (type) context.unificationTable.addConstraint(ast.id, { exactly: type });
      else context.names.set(name, { variable: ast.id });

      return type ?? { variable: ast.id };
    }
    case NodeType.SCRIPT:
    case NodeType.SEQUENCE: {
      const types = ast.children.map((child) => infer(child, context));
      context.unificationTable.addConstraint(ast.id, { equals: ast.children[types.length - 1].id });
      return types[types.length - 1];
    }
    case NodeType.DELIMITED_APPLICATION:
    case NodeType.APPLICATION: {
      infer(ast.children[0], context);
      const arg_type = infer(ast.children[1], context);
      const return_type: Type = { variable: ast.id };

      context.unificationTable.addConstraint(ast.id, { exactly: return_type });
      context.unificationTable.addConstraint(ast.children[0].id, {
        exactly: { fn: { arg: arg_type, return: return_type } },
      });

      return return_type;
    }
    case NodeType.FUNCTION: {
      return context.bind(ast.children[0], (argType) => {
        const returnType = infer(ast.children[1], context);
        const type = { fn: { arg: argType, return: returnType } };
        context.unificationTable.addConstraint(ast.id, { exactly: type });
        return type;
      });
    }
    case NodeType.ADD: {
      context.unificationTable.addConstraint(ast.id, { exactly: "int" });
      return "int";
    }
    case NodeType.MODULE:
    default:
      unreachable(`cant infer top types ${ast.type}`);
  }
};

const constrain = (ast: Tree, context: Context, expectedType: Type): void => {};

export const inferPhysical = (typeMap: TypeSchema): PhysicalTypeSchema => {
  const physicalTypeMap: PhysicalTypeSchema = new Map();

  const translate = (type: Type, id?: number): PhysicalType => {
    if (type === "int") return { int: 32 };
    if (type === "float") return { float: 32 };
    if (type === "string") return { pointer: { int: 8 } };
    if (type === "void") return "void";
    if (type === "unknown") return "unknown";
    assert(typeof type === "object");

    // must be resolved by that point
    assert(!("variable" in type));
    assert(!("and" in type));

    if ("fn" in type) {
      const arg = translate(type.fn.arg);
      const returnType = translate(type.fn.return);
      const freeVars = (id && inject(Injectable.ClosureVariablesMap).get(id)) || [];
      assert(freeVars);
      const closure = freeVars.map((name) => {
        const type = typeMap.get(name)!;
        const physicalType = translate(type, name);
        physicalTypeMap.set(name, physicalType);
        return physicalType;
      });

      return { fn: { args: [arg], ret: returnType, closure } };
    }
    if ("atom" in type) return { int: 32 };

    unreachable("cant convert type to LLVM type");
  };

  typeMap.forEach((type, id) => {
    if (physicalTypeMap.has(id)) return;
    physicalTypeMap.set(id, translate(type, id));
  });

  return physicalTypeMap;
};

export const substituteConstraints = (ast: Tree, context: Context, map: TypeSchema = new Map()): TypeSchema => {
  if (map.has(ast.id)) return map;
  switch (ast.type) {
    case NodeType.FUNCTION: {
      substituteConstraints(ast.children[1], context, map);
      map.set(ast.id, context.unificationTable.resolveType(ast.id));
      return map;
    }
    default: {
      ast.children.forEach((child) => substituteConstraints(child, context, map));
      map.set(ast.id, context.unificationTable.resolveType(ast.id));
      return map;
    }
  }
};

export const inferTypes = (ast: Tree): TypeSchema => {
  const context = new Context();
  resolve(ast, globalResolvedNamesArray);
  infer(ast, context);
  context.unificationTable.truncateTautologies();
  return substituteConstraints(ast, context);
  // console.dir(
  //   {
  //     // ast,
  //     table: context.unificationTable,
  //     // types: inject(Injectable.TypeMap),
  //     closures: inject(Injectable.ClosureVariablesMap),
  //     bound: inject(Injectable.BoundVariablesMap),
  //     nodeToVariable: inject(Injectable.NodeToVariableMap),
  //   },
  //   { depth: null }
  // );
  // inferPhysical();
};
