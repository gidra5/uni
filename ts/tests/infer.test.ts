import { beforeEach, describe, expect, test } from "vitest";
import { Tree } from "../src/ast";
import { parseScript } from "../src/parser/parser";
import { parseTokenGroups } from "../src/parser/tokenGroups";
import { Injectable, register } from "../src/utils/injector";
import { FileMap } from "codespan-napi";
import { inferTypes } from "../src/analysis/types/infer";
import { desugar } from "../src/analysis/desugar";

beforeEach(() => {
  register(Injectable.FileMap, new FileMap());
  register(Injectable.NextId, 0);
  register(Injectable.ASTNodePrecedenceMap, new Map());
  register(Injectable.PositionMap, new Map());
});

function clearIds(ast: Tree) {
  if (ast.children.length > 0) {
    ast.children.forEach(clearIds);
  }
  delete (ast as any).id;
  return ast;
}

const testCase = async (input: string) => {
  const tokens = parseTokenGroups(input);
  const ast = parseScript(tokens);
  const desugared = desugar(ast);
  inferTypes(desugared);

  // expect(ast).toMatchSnapshot();
  expect(clearIds(ast)).toMatchSnapshot();
};

describe("compilation", () => {
  test("fix", async () => await testCase(`fn f -> (fn x -> x x) (fn x -> f(x x))`));
  test("either", async () => await testCase(`(((fn x -> fn m -> fn n -> m x) 1) fn x -> x) fn x -> x`));
  test("apply", async () => await testCase(`((fn f -> fn x -> f x) fn x -> x) 2`));
  test("wrapper", async () => await testCase(`((fn x -> fn m -> m x) 2) fn x -> x`));
  test("church tuple", async () => await testCase(`((fn x -> fn y -> fn m -> m x y) 1 2) fn x -> fn _ -> x`));
  test("function closure", async () => await testCase(`(fn x -> fn y -> y + 2 * x) 1 2`));
  test("function deep closure", async () => await testCase(`(fn x -> fn y -> fn z -> x + y + z) 1 3 5`));
  test.only("function application and literal", async () => await testCase(`(fn x -> x + x) 2`));
  test("float", async () => await testCase(`1.1`));
  test("int", async () => await testCase(`1`));
  test("sequence", async () => await testCase(`1; "hello world 2!"`));
  test("hello world string", async () => await testCase(`"hello world!"`));
  test.skip("print 1", async () => await testCase(`print 1`));
  test.skip("print 2", async () => await testCase(`print "x"`));
  test.skip("print 3", async () => await testCase(`print 1; print "x"`));
  test.skip("function application and literal print", async () => await testCase(`print((fn x -> x + x) 2)`));
});
