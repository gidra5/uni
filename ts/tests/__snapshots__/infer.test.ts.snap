// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`compilation > apply 1`] = `
{
  "children": [
    {
      "children": [
        {
          "children": [
            {
              "children": [
                {
                  "children": [],
                  "data": {
                    "value": "f",
                  },
                  "type": "name",
                },
                {
                  "children": [
                    {
                      "children": [],
                      "data": {
                        "value": "x",
                      },
                      "type": "name",
                    },
                    {
                      "children": [
                        {
                          "children": [],
                          "data": {
                            "type": {
                              "fn": {
                                "arg": "int",
                                "return": "int",
                              },
                            },
                            "value": "f",
                          },
                          "type": "name",
                        },
                        {
                          "children": [],
                          "data": {
                            "type": "int",
                            "value": "x",
                          },
                          "type": "name",
                        },
                      ],
                      "data": {
                        "type": "int",
                      },
                      "type": "application",
                    },
                  ],
                  "data": {
                    "isTopFunction": true,
                    "type": {
                      "fn": {
                        "arg": "int",
                        "return": "int",
                      },
                    },
                  },
                  "type": "func",
                },
              ],
              "data": {
                "isTopFunction": true,
                "type": {
                  "fn": {
                    "arg": {
                      "fn": {
                        "arg": "int",
                        "return": "int",
                      },
                    },
                    "return": {
                      "fn": {
                        "arg": "int",
                        "return": "int",
                      },
                    },
                  },
                },
              },
              "type": "func",
            },
            {
              "children": [
                {
                  "children": [],
                  "data": {
                    "value": "x",
                  },
                  "type": "name",
                },
                {
                  "children": [],
                  "data": {
                    "type": "int",
                    "value": "x",
                  },
                  "type": "name",
                },
              ],
              "data": {
                "isTopFunction": true,
                "type": {
                  "fn": {
                    "arg": "int",
                    "return": "int",
                  },
                },
              },
              "type": "func",
            },
          ],
          "data": {
            "type": {
              "fn": {
                "arg": "int",
                "return": "int",
              },
            },
          },
          "type": "application",
        },
        {
          "children": [],
          "data": {
            "type": "int",
            "value": 2,
          },
          "type": "number",
        },
      ],
      "data": {
        "type": "int",
      },
      "type": "application",
    },
  ],
  "data": {
    "type": "int",
  },
  "type": "script",
}
`;

exports[`compilation > church tuple 1`] = `
{
  "children": [
    {
      "children": [
        {
          "children": [
            {
              "children": [
                {
                  "children": [
                    {
                      "children": [],
                      "data": {
                        "value": "x",
                      },
                      "type": "name",
                    },
                    {
                      "children": [
                        {
                          "children": [],
                          "data": {
                            "value": "y",
                          },
                          "type": "name",
                        },
                        {
                          "children": [
                            {
                              "children": [],
                              "data": {
                                "value": "m",
                              },
                              "type": "name",
                            },
                            {
                              "children": [
                                {
                                  "children": [
                                    {
                                      "children": [],
                                      "data": {
                                        "type": {
                                          "fn": {
                                            "arg": "int",
                                            "return": {
                                              "fn": {
                                                "arg": "int",
                                                "return": "int",
                                              },
                                            },
                                          },
                                        },
                                        "value": "m",
                                      },
                                      "type": "name",
                                    },
                                    {
                                      "children": [],
                                      "data": {
                                        "type": "int",
                                        "value": "x",
                                      },
                                      "type": "name",
                                    },
                                  ],
                                  "data": {
                                    "type": {
                                      "fn": {
                                        "arg": "int",
                                        "return": "int",
                                      },
                                    },
                                  },
                                  "type": "application",
                                },
                                {
                                  "children": [],
                                  "data": {
                                    "type": "int",
                                    "value": "y",
                                  },
                                  "type": "name",
                                },
                              ],
                              "data": {
                                "type": "int",
                              },
                              "type": "application",
                            },
                          ],
                          "data": {
                            "isTopFunction": true,
                            "type": {
                              "fn": {
                                "arg": {
                                  "fn": {
                                    "arg": "int",
                                    "return": {
                                      "fn": {
                                        "arg": "int",
                                        "return": "int",
                                      },
                                    },
                                  },
                                },
                                "return": "int",
                              },
                            },
                          },
                          "type": "func",
                        },
                      ],
                      "data": {
                        "isTopFunction": true,
                        "type": {
                          "fn": {
                            "arg": "int",
                            "return": {
                              "fn": {
                                "arg": {
                                  "fn": {
                                    "arg": "int",
                                    "return": {
                                      "fn": {
                                        "arg": "int",
                                        "return": "int",
                                      },
                                    },
                                  },
                                },
                                "return": "int",
                              },
                            },
                          },
                        },
                      },
                      "type": "func",
                    },
                  ],
                  "data": {
                    "isTopFunction": true,
                    "type": {
                      "fn": {
                        "arg": "int",
                        "return": {
                          "fn": {
                            "arg": "int",
                            "return": {
                              "fn": {
                                "arg": {
                                  "fn": {
                                    "arg": "int",
                                    "return": {
                                      "fn": {
                                        "arg": "int",
                                        "return": "int",
                                      },
                                    },
                                  },
                                },
                                "return": "int",
                              },
                            },
                          },
                        },
                      },
                    },
                  },
                  "type": "func",
                },
                {
                  "children": [],
                  "data": {
                    "type": "int",
                    "value": 1,
                  },
                  "type": "number",
                },
              ],
              "data": {
                "type": {
                  "fn": {
                    "arg": "int",
                    "return": {
                      "fn": {
                        "arg": {
                          "fn": {
                            "arg": "int",
                            "return": {
                              "fn": {
                                "arg": "int",
                                "return": "int",
                              },
                            },
                          },
                        },
                        "return": "int",
                      },
                    },
                  },
                },
              },
              "type": "application",
            },
            {
              "children": [],
              "data": {
                "type": "int",
                "value": 2,
              },
              "type": "number",
            },
          ],
          "data": {
            "type": {
              "fn": {
                "arg": {
                  "fn": {
                    "arg": "int",
                    "return": {
                      "fn": {
                        "arg": "int",
                        "return": "int",
                      },
                    },
                  },
                },
                "return": "int",
              },
            },
          },
          "type": "application",
        },
        {
          "children": [
            {
              "children": [],
              "data": {
                "value": "x",
              },
              "type": "name",
            },
            {
              "children": [
                {
                  "children": [],
                  "data": {},
                  "type": "placeholder",
                },
                {
                  "children": [],
                  "data": {
                    "type": "int",
                    "value": "x",
                  },
                  "type": "name",
                },
              ],
              "data": {
                "isTopFunction": true,
                "type": {
                  "fn": {
                    "arg": "int",
                    "return": "int",
                  },
                },
              },
              "type": "func",
            },
          ],
          "data": {
            "isTopFunction": true,
            "type": {
              "fn": {
                "arg": "int",
                "return": {
                  "fn": {
                    "arg": "int",
                    "return": "int",
                  },
                },
              },
            },
          },
          "type": "func",
        },
      ],
      "data": {
        "type": "int",
      },
      "type": "application",
    },
  ],
  "data": {
    "type": "int",
  },
  "type": "script",
}
`;

exports[`compilation > either 1`] = `
{
  "children": [
    {
      "children": [
        {
          "children": [
            {
              "children": [
                {
                  "children": [
                    {
                      "children": [],
                      "data": {
                        "value": "x",
                      },
                      "type": "name",
                    },
                    {
                      "children": [
                        {
                          "children": [],
                          "data": {
                            "value": "m",
                          },
                          "type": "name",
                        },
                        {
                          "children": [
                            {
                              "children": [],
                              "data": {
                                "value": "n",
                              },
                              "type": "name",
                            },
                            {
                              "children": [
                                {
                                  "children": [],
                                  "data": {
                                    "type": {
                                      "fn": {
                                        "arg": "int",
                                        "return": "int",
                                      },
                                    },
                                    "value": "m",
                                  },
                                  "type": "name",
                                },
                                {
                                  "children": [],
                                  "data": {
                                    "type": "int",
                                    "value": "x",
                                  },
                                  "type": "name",
                                },
                              ],
                              "data": {
                                "type": "int",
                              },
                              "type": "application",
                            },
                          ],
                          "data": {
                            "isTopFunction": true,
                            "type": {
                              "fn": {
                                "arg": {
                                  "fn": {
                                    "arg": "unknown",
                                    "return": "unknown",
                                  },
                                },
                                "return": "int",
                              },
                            },
                          },
                          "type": "func",
                        },
                      ],
                      "data": {
                        "isTopFunction": true,
                        "type": {
                          "fn": {
                            "arg": {
                              "fn": {
                                "arg": "int",
                                "return": "int",
                              },
                            },
                            "return": {
                              "fn": {
                                "arg": {
                                  "fn": {
                                    "arg": "unknown",
                                    "return": "unknown",
                                  },
                                },
                                "return": "int",
                              },
                            },
                          },
                        },
                      },
                      "type": "func",
                    },
                  ],
                  "data": {
                    "isTopFunction": true,
                    "type": {
                      "fn": {
                        "arg": "int",
                        "return": {
                          "fn": {
                            "arg": {
                              "fn": {
                                "arg": "int",
                                "return": "int",
                              },
                            },
                            "return": {
                              "fn": {
                                "arg": {
                                  "fn": {
                                    "arg": "unknown",
                                    "return": "unknown",
                                  },
                                },
                                "return": "int",
                              },
                            },
                          },
                        },
                      },
                    },
                  },
                  "type": "func",
                },
                {
                  "children": [],
                  "data": {
                    "type": "int",
                    "value": 1,
                  },
                  "type": "number",
                },
              ],
              "data": {
                "type": {
                  "fn": {
                    "arg": {
                      "fn": {
                        "arg": "int",
                        "return": "int",
                      },
                    },
                    "return": {
                      "fn": {
                        "arg": {
                          "fn": {
                            "arg": "unknown",
                            "return": "unknown",
                          },
                        },
                        "return": "int",
                      },
                    },
                  },
                },
              },
              "type": "application",
            },
            {
              "children": [
                {
                  "children": [],
                  "data": {
                    "value": "x",
                  },
                  "type": "name",
                },
                {
                  "children": [],
                  "data": {
                    "type": "int",
                    "value": "x",
                  },
                  "type": "name",
                },
              ],
              "data": {
                "isTopFunction": true,
                "type": {
                  "fn": {
                    "arg": "int",
                    "return": "int",
                  },
                },
              },
              "type": "func",
            },
          ],
          "data": {
            "type": {
              "fn": {
                "arg": {
                  "fn": {
                    "arg": "unknown",
                    "return": "unknown",
                  },
                },
                "return": "int",
              },
            },
          },
          "type": "application",
        },
        {
          "children": [
            {
              "children": [],
              "data": {
                "value": "x",
              },
              "type": "name",
            },
            {
              "children": [],
              "data": {
                "type": "unknown",
                "value": "x",
              },
              "type": "name",
            },
          ],
          "data": {
            "isTopFunction": true,
            "type": {
              "fn": {
                "arg": "unknown",
                "return": "unknown",
              },
            },
          },
          "type": "func",
        },
      ],
      "data": {
        "type": "int",
      },
      "type": "application",
    },
  ],
  "data": {
    "type": "int",
  },
  "type": "script",
}
`;

exports[`compilation > float 1`] = `
{
  "children": [
    {
      "children": [],
      "data": {
        "type": "float",
        "value": 1.1,
      },
      "type": "number",
    },
  ],
  "data": {
    "type": "float",
  },
  "type": "script",
}
`;

exports[`compilation > function application and literal 1`] = `
{
  "children": [
    {
      "children": [
        {
          "children": [
            {
              "children": [],
              "data": {
                "value": "x",
              },
              "type": "name",
            },
            {
              "children": [
                {
                  "children": [],
                  "data": {
                    "type": "int",
                    "value": "x",
                  },
                  "type": "name",
                },
                {
                  "children": [],
                  "data": {
                    "type": "int",
                    "value": "x",
                  },
                  "type": "name",
                },
              ],
              "data": {
                "type": "int",
              },
              "type": "add",
            },
          ],
          "data": {
            "isTopFunction": true,
            "type": {
              "fn": {
                "arg": "int",
                "return": "int",
              },
            },
          },
          "type": "func",
        },
        {
          "children": [],
          "data": {
            "type": "int",
            "value": 2,
          },
          "type": "number",
        },
      ],
      "data": {
        "type": "int",
      },
      "type": "application",
    },
  ],
  "data": {
    "type": "int",
  },
  "type": "script",
}
`;

exports[`compilation > function closure 1`] = `
{
  "children": [
    {
      "children": [
        {
          "children": [
            {
              "children": [
                {
                  "children": [],
                  "data": {
                    "value": "x",
                  },
                  "type": "name",
                },
                {
                  "children": [
                    {
                      "children": [],
                      "data": {
                        "value": "y",
                      },
                      "type": "name",
                    },
                    {
                      "children": [
                        {
                          "children": [],
                          "data": {
                            "type": "int",
                            "value": "y",
                          },
                          "type": "name",
                        },
                        {
                          "children": [
                            {
                              "children": [],
                              "data": {
                                "type": "int",
                                "value": 2,
                              },
                              "type": "number",
                            },
                            {
                              "children": [],
                              "data": {
                                "type": "int",
                                "value": "x",
                              },
                              "type": "name",
                            },
                          ],
                          "data": {
                            "type": "int",
                          },
                          "type": "*",
                        },
                      ],
                      "data": {
                        "type": "int",
                      },
                      "type": "add",
                    },
                  ],
                  "data": {
                    "isTopFunction": true,
                    "type": {
                      "fn": {
                        "arg": "int",
                        "return": "int",
                      },
                    },
                  },
                  "type": "func",
                },
              ],
              "data": {
                "isTopFunction": true,
                "type": {
                  "fn": {
                    "arg": "int",
                    "return": {
                      "fn": {
                        "arg": "int",
                        "return": "int",
                      },
                    },
                  },
                },
              },
              "type": "func",
            },
            {
              "children": [],
              "data": {
                "type": "int",
                "value": 1,
              },
              "type": "number",
            },
          ],
          "data": {
            "type": {
              "fn": {
                "arg": "int",
                "return": "int",
              },
            },
          },
          "type": "application",
        },
        {
          "children": [],
          "data": {
            "type": "int",
            "value": 2,
          },
          "type": "number",
        },
      ],
      "data": {
        "type": "int",
      },
      "type": "application",
    },
  ],
  "data": {
    "type": "int",
  },
  "type": "script",
}
`;

exports[`compilation > function deep closure 1`] = `
{
  "children": [
    {
      "children": [
        {
          "children": [
            {
              "children": [
                {
                  "children": [
                    {
                      "children": [],
                      "data": {
                        "value": "x",
                      },
                      "type": "name",
                    },
                    {
                      "children": [
                        {
                          "children": [],
                          "data": {
                            "value": "y",
                          },
                          "type": "name",
                        },
                        {
                          "children": [
                            {
                              "children": [],
                              "data": {
                                "value": "z",
                              },
                              "type": "name",
                            },
                            {
                              "children": [
                                {
                                  "children": [],
                                  "data": {
                                    "type": "int",
                                    "value": "x",
                                  },
                                  "type": "name",
                                },
                                {
                                  "children": [],
                                  "data": {
                                    "type": "int",
                                    "value": "y",
                                  },
                                  "type": "name",
                                },
                                {
                                  "children": [],
                                  "data": {
                                    "type": "int",
                                    "value": "z",
                                  },
                                  "type": "name",
                                },
                              ],
                              "data": {
                                "type": "int",
                              },
                              "type": "add",
                            },
                          ],
                          "data": {
                            "isTopFunction": true,
                            "type": {
                              "fn": {
                                "arg": "int",
                                "return": "int",
                              },
                            },
                          },
                          "type": "func",
                        },
                      ],
                      "data": {
                        "isTopFunction": true,
                        "type": {
                          "fn": {
                            "arg": "int",
                            "return": {
                              "fn": {
                                "arg": "int",
                                "return": "int",
                              },
                            },
                          },
                        },
                      },
                      "type": "func",
                    },
                  ],
                  "data": {
                    "isTopFunction": true,
                    "type": {
                      "fn": {
                        "arg": "int",
                        "return": {
                          "fn": {
                            "arg": "int",
                            "return": {
                              "fn": {
                                "arg": "int",
                                "return": "int",
                              },
                            },
                          },
                        },
                      },
                    },
                  },
                  "type": "func",
                },
                {
                  "children": [],
                  "data": {
                    "type": "int",
                    "value": 1,
                  },
                  "type": "number",
                },
              ],
              "data": {
                "type": {
                  "fn": {
                    "arg": "int",
                    "return": {
                      "fn": {
                        "arg": "int",
                        "return": "int",
                      },
                    },
                  },
                },
              },
              "type": "application",
            },
            {
              "children": [],
              "data": {
                "type": "int",
                "value": 3,
              },
              "type": "number",
            },
          ],
          "data": {
            "type": {
              "fn": {
                "arg": "int",
                "return": "int",
              },
            },
          },
          "type": "application",
        },
        {
          "children": [],
          "data": {
            "type": "int",
            "value": 5,
          },
          "type": "number",
        },
      ],
      "data": {
        "type": "int",
      },
      "type": "application",
    },
  ],
  "data": {
    "type": "int",
  },
  "type": "script",
}
`;

exports[`compilation > hello world string 1`] = `
{
  "children": [
    {
      "children": [],
      "data": {
        "type": "string",
        "value": "hello world!",
      },
      "type": "string",
    },
  ],
  "data": {
    "type": "string",
  },
  "type": "script",
}
`;

exports[`compilation > int 1`] = `
{
  "children": [
    {
      "children": [],
      "data": {
        "type": "int",
        "value": 1,
      },
      "type": "number",
    },
  ],
  "data": {
    "type": "int",
  },
  "type": "script",
}
`;

exports[`compilation > print 1 1`] = `
{
  "children": [
    {
      "children": [
        {
          "children": [],
          "data": {
            "type": "void",
            "value": "print",
          },
          "type": "name",
        },
        {
          "children": [],
          "data": {
            "type": "int",
            "value": 1,
          },
          "type": "number",
        },
      ],
      "data": {
        "type": "unknown",
      },
      "type": "application",
    },
  ],
  "data": {
    "type": "unknown",
  },
  "type": "script",
}
`;

exports[`compilation > two prints 1`] = `
{
  "children": [
    {
      "children": [],
      "data": {
        "type": "string",
        "value": "hello world!",
      },
      "type": "string",
    },
    {
      "children": [],
      "data": {
        "type": "string",
        "value": "hello world 2!",
      },
      "type": "string",
    },
  ],
  "data": {
    "type": "string",
  },
  "type": "script",
}
`;

exports[`compilation > wrapper 1`] = `
{
  "children": [
    {
      "children": [
        {
          "children": [
            {
              "children": [
                {
                  "children": [],
                  "data": {
                    "value": "x",
                  },
                  "type": "name",
                },
                {
                  "children": [
                    {
                      "children": [],
                      "data": {
                        "value": "m",
                      },
                      "type": "name",
                    },
                    {
                      "children": [
                        {
                          "children": [],
                          "data": {
                            "type": {
                              "fn": {
                                "arg": "int",
                                "return": "int",
                              },
                            },
                            "value": "m",
                          },
                          "type": "name",
                        },
                        {
                          "children": [],
                          "data": {
                            "type": "int",
                            "value": "x",
                          },
                          "type": "name",
                        },
                      ],
                      "data": {
                        "type": "int",
                      },
                      "type": "application",
                    },
                  ],
                  "data": {
                    "isTopFunction": true,
                    "type": {
                      "fn": {
                        "arg": {
                          "fn": {
                            "arg": "int",
                            "return": "int",
                          },
                        },
                        "return": "int",
                      },
                    },
                  },
                  "type": "func",
                },
              ],
              "data": {
                "isTopFunction": true,
                "type": {
                  "fn": {
                    "arg": "int",
                    "return": {
                      "fn": {
                        "arg": {
                          "fn": {
                            "arg": "int",
                            "return": "int",
                          },
                        },
                        "return": "int",
                      },
                    },
                  },
                },
              },
              "type": "func",
            },
            {
              "children": [],
              "data": {
                "type": "int",
                "value": 2,
              },
              "type": "number",
            },
          ],
          "data": {
            "type": {
              "fn": {
                "arg": {
                  "fn": {
                    "arg": "int",
                    "return": "int",
                  },
                },
                "return": "int",
              },
            },
          },
          "type": "application",
        },
        {
          "children": [
            {
              "children": [],
              "data": {
                "value": "x",
              },
              "type": "name",
            },
            {
              "children": [],
              "data": {
                "type": "int",
                "value": "x",
              },
              "type": "name",
            },
          ],
          "data": {
            "isTopFunction": true,
            "type": {
              "fn": {
                "arg": "int",
                "return": "int",
              },
            },
          },
          "type": "func",
        },
      ],
      "data": {
        "type": "int",
      },
      "type": "application",
    },
  ],
  "data": {
    "type": "int",
  },
  "type": "script",
}
`;
